openhtf/__init__.py

```diff
@@ -71,21 +72,39 @@ def PhaseTwo(test):
       # Analyze widget integration status
 
     Test(PhaseOne, PhaseTwo).Execute()
+
+  Note that Test() objects *must* be created in the main thread, but can be
+  .Execute()'d in a separate thread.
   """
 
   def __init__(self, *phases, **metadata):
     code_info = test_record.CodeInfo.ForModuleFromStack(levels_up=2)
+    self._test_options = TestOptions()
     self._test_info = TestData(phases, metadata=metadata, code_info=code_info)
-    self._output_callbacks = []
-    self.loop = False
+    self._lock = threading.Lock()
+    self._stopped = False
+    # Make sure Configure() gets called at least once before Execute().  The
+    # user might call Configure() again to override options, but we don't want
+    # to force them to if they want to use defaults.  For default values, see
+    # the class definition of TestOptions.
+    self.Configure()
+    # TODO(madsci): Fix this to play nice with multiple Test instances.
+    signal.signal(signal.SIGINT, self.Stop)
 
   def AddOutputCallback(self, callback):
+    """DEPRECATED: Use AddOutputCallbacks() instead."""
+    # TODO(madsci): Remove this before we push to PyPI, here for transitionary
+    # purposes. 
+    raise AttributeError(
+        'DEPRECATED, use AddOutputCallbacks() instead of AddOutputCallback()')
+
+  def AddOutputCallbacks(self, *callbacks):
     """Add the given function as an output module to this test."""
-    self._output_callbacks.append(callback)
+    self._test_options.output_callbacks.extend(callbacks)
```

Comment:
The latter is preferred, and that's documented in the docstring for `Configure()`.  I do agree that generally we want to avoid having multiple ways to do things, but any setter method in Python could really just poke at `foo._thing` directly rather than call `foo.SetThing` - that's similar to this case, `Configure()` only exposes output_callbacks by virtue of those callbacks being stored in the same manner as other options, but you're really not supposed to access them that way.

short answer, don't do the former, there is only one accepted way to do it, it's the latter :)